// coc-snippets/lib/list/snippet.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/******************************************************************
MIT License http://www.opensource.org/licenses/mit-license.php
Author Qiming Zhao <chemzqm@gmail> (https://github.com/chemzqm)
*******************************************************************/
const coc_nvim_1 = require("coc.nvim");
const os_1 = __importDefault(require("os"));
class SnippetsList extends coc_nvim_1.BasicList {
    constructor(nvim, manager) {
        super(nvim);
        this.manager = manager;
        this.name = 'snippets';
        this.description = 'snippets list';
        this.addLocationActions();
    }
    async loadItems(context) {
        let { window } = context;
        let valid = await window.valid;
        if (!valid)
            return;
        let buf = await window.buffer;
        let doc = coc_nvim_1.workspace.getDocument(buf.id);
        if (!doc)
            return [];
        let snippets = await this.manager.getSnippets(doc.filetype);
        let res = [];
        for (let snip of snippets) {
            let pos = coc_nvim_1.Position.create(snip.lnum, 0);
            let location = coc_nvim_1.Location.create(coc_nvim_1.Uri.file(snip.filepath).toString(), coc_nvim_1.Range.create(pos, coc_nvim_1.Position.create(snip.lnum, 1)));
            let prefix = snip.prefix;
            if (prefix.length < 20) {
                prefix = `${prefix}${' '.repeat(20 - prefix.length)}`;
            }
            res.push({
                label: `${prefix}\t${snip.description}\t${snip.filepath.replace(os_1.default.homedir(), '~')}`,
                filterText: `${snip.prefix} ${snip.description}`,
                location,
                data: { prefix: snip.prefix }
            });
        }
        return res;
    }
    async doHighlight() {
        let { nvim } = coc_nvim_1.workspace;
        nvim.pauseNotification();
        nvim.command('syntax match CocSnippetsPrefix /\\v^[^\\t]+/ contained containedin=CocSnippetsLine', true);
        nvim.command('syntax match CocSnippetsFile /\\v\\t\\S+$/ contained containedin=CocSnippetsLine', true);
        nvim.command('highlight default link CocSnippetsPrefix Identifier', true);
        nvim.command('highlight default link CocSnippetsFile Comment', true);
        void nvim.resumeNotification(false, true);
    }
}
exports.default = SnippetsList;
